### PR TITLE
Read-Only Replica: Implement common base classes

### PR DESCRIPTION
**Pull request by toly-kournik**

### CODE REVIEW

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 1]
```diff
@@ -12,160 +12,111 @@
...
-    onMessage(validMsg);
-  } else {
-    onReportAboutInvalidMessage(msg);
-    delete msg;
-    return;
-  }
-}
-
-template <typename T>
-void ReplicaImp::messageHandlerWithIgnoreLogic(MessageBase *msg) {
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicaImp.cpp]: Why did you remove messageHandlerWithIgnoreLogic ?

File: bftengine/src/bftengine/messages/PrePrepareMsg.cpp [hunk 2]
```diff
@@ -40,55 +40,38 @@ PrePrepareMsg* PrePrepareMsg::createNullPrePrepareMsg(ReplicaId sender, ViewNum
...
-  // check seqNum
-  if (tmp->b()->seqNum == 0) return false;
-
   // check flags
-  const uint16_t flags = tmp->b()->flags;
+  const uint16_t flags = b()->flags;
   const bool isNull = ((flags & 0x1) == 0);
   const bool isReady = (((flags >> 1) & 0x1) == 1);
-  const uint16_t firstPath = ((flags >> 2) & 0x3);
+  const uint16_t firstPath_ = ((flags >> 2) & 0x3);
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/messages/PrePrepareMsg.cpp]: there's a function with a name firstPath()

File: bftengine/include/bftengine/Replica.hpp [hunk 3]
```diff
@@ -43,24 +43,23 @@ class RequestsHandler {
...
-                                   ICommunication *communication,
-                                   MetadataStorage *metadataStorage);
+  static IReplica *createNewReplica(ReplicaConfig *replicaConfig,
+                                    IRequestsHandler *requestsHandler,
+                                    IStateTransfer *stateTransfer,
+                                    ICommunication *communication,
+                                    MetadataStorage *metadataStorage);
 
-  virtual ~Replica() = 0;
+  virtual ~IReplica(){};
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/include/bftengine/Replica.hpp]: done

File: bftengine/src/bftengine/BFTEngine.cpp [hunk 4]
```diff
@@ -29,45 +31,32 @@ bool cryptoInitialized = false;
 std::mutex mutexForCryptoInitialization;
 }  // namespace
 
-class ReplicaInternal : public Replica {
+class ReplicaInternal : public IReplica {
+  friend class IReplica;
+
  public:
-  ~ReplicaInternal() override;
+  virtual ~ReplicaInternal() {}
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/BFTEngine.cpp]: removed this empty destructor

File: bftengine/src/bftengine/ReplicaForStateTransfer.hpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+ public:
+  ReplicaForStateTransfer(const ReplicaConfig&,
+                          IStateTransfer*,
+                          std::shared_ptr<MsgsCommunicator>,
+                          std::shared_ptr<MsgHandlersRegistrator>,
+                          bool firstTime  // TODO [TK] get rid of this
+  );
+
+  virtual ~ReplicaForStateTransfer() {}
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/ReplicaForStateTransfer.hpp]: done

File: bftengine/include/bftengine/Replica.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/include/bftengine/Replica.hpp]: Pure virtual destructor still needs to be defined and only makes the class abstract, which it's already. 
So I followed Yulia's comment and made it default

File: bftengine/src/bftengine/BFTEngine.cpp [hunk 6]
```diff
@@ -29,45 +31,32 @@ bool cryptoInitialized = false;
...
 
   bool isRunning() const override;
 
-  uint64_t getLastExecutedSequenceNum() const override;
-
-  bool requestsExecutionWasInterrupted() const override;
+  int64_t getLastExecutedSequenceNum() const override { return replica_->getLastExecutedSequenceNum(); }
 
-  void start() override;
+  virtual void start() override;
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/BFTEngine.cpp]: fixed

File: bftengine/src/bftengine/BFTEngine.cpp [hunk 7]
```diff
@@ -29,45 +31,32 @@ bool cryptoInitialized = false;
 std::mutex mutexForCryptoInitialization;
 }  // namespace
 
-class ReplicaInternal : public Replica {
+class ReplicaInternal : public IReplica {
+  friend class IReplica;
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/BFTEngine.cpp]: IReplica::createNewReplica() uses private members

File: bftengine/src/bftengine/BFTEngine.cpp [hunk 8]
```diff
@@ -94,33 +83,36 @@ void ReplicaInternal::restartForDebug(uint32_t delayMillis) {
...
 
 }  // namespace bftEngine::impl
 
 namespace bftEngine {
-Replica *Replica::createNewReplica(ReplicaConfig *replicaConfig,
-                                   RequestsHandler *requestsHandler,
-                                   IStateTransfer *stateTransfer,
-                                   ICommunication *communication,
-                                   MetadataStorage *metadataStorage) {
+/**
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/BFTEngine.cpp]: fixed

File: bftengine/src/bftengine/SimpleClientImp.cpp [hunk 9]
```diff
@@ -104,11 +104,9 @@ class SimpleClientImp : public SimpleClient, public IReceiver {
...
 
 void SimpleClientImp::onMessageFromReplica(MessageBase* msg) {
-  ClientReplyMsg* replyMsg = nullptr;
-  if (!ClientReplyMsg::ToActualMsgType(clientId_, msg, replyMsg)) {
-    delete msg;
-    return;
-  }
+  ClientReplyMsg* replyMsg = static_cast<ClientReplyMsg*>(msg);
+  ReplicasInfo dummy(0, 4, 1, 0, false, false);
+  replyMsg->validate(dummy);
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/SimpleClientImp.cpp]: You are right, but it's a common API for all external messages

File: bftengine/include/bftengine/IStateTransfer.hpp [hunk 10]
```diff
@@ -82,6 +82,8 @@ class IReplicaForStateTransfer {
   // the timer is disabled when timerPeriodMilli==0
   // (notice that the state transfer module can use its own timers and threads)
   virtual void changeStateTransferTimerPeriod(uint32_t timerPeriodMilli) = 0;
+
+  virtual ~IReplicaForStateTransfer() {}
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/include/bftengine/IStateTransfer.hpp]: removed

File: bftengine/src/bftengine/ReplicaBase.cpp [hunk 11]
```diff
@@ -0,0 +1,74 @@
+// Concord
+//
+// Copyright (c) 2018, 2019 VMware, Inc. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/ReplicaBase.cpp]: Let's do it in a separate PR for all the codebase

File: bftengine/src/bftengine/ReplicaImp.hpp [hunk 12]
```diff
@@ -54,26 +42,22 @@ class ReqMissingDataMsg;
...
 class PrepareFullMsg;
 class SimpleAckMsg;
-class StateTransferMsg;
 class ReplicaStatusMsg;
 class ReplicaImp;
+class LoadedReplicaData;
+class PersistentStorage;
 
 using bftEngine::ReplicaConfig;
+using std::shared_ptr;
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/ReplicaImp.hpp]: shared_ptr was used without a namespace. using std::shared_ptr was declared in other include file.

File: bftengine/src/bftengine/ReplicaForStateTransfer.hpp [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+#include "ReplicaBase.hpp"
+#include "IStateTransfer.hpp"
+
+namespace bftEngine::impl {
+
+/**
+ *
+ */
+class ReplicaForStateTransfer : public IReplicaForStateTransfer, public ReplicaBase {
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/ReplicaForStateTransfer.hpp]: So we'll leave refactoring for the future.
Also let's do a separate PR for different renamings.
There are a lot of changes in this PR and it took already too many cycles - I want to merge it.

File: bftengine/src/bftengine/BFTEngine.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (andrewjstone) [reply bftengine/src/bftengine/BFTEngine.cpp]: Ok for now.

File: bftengine/src/bftengine/ReplicaImp.hpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (andrewjstone) [reply bftengine/src/bftengine/ReplicaImp.hpp]: ok.

File: bftengine/src/bftengine/ReplicaBase.hpp [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+#include "PrimitiveTypes.hpp"
+#include "ReplicaConfig.hpp"
+#include "SeqNumInfo.hpp"
+#include "DebugStatistics.hpp"
+#include "Metrics.hpp"
+#include "Timers.hpp"
+
+namespace bftEngine::impl {
+
+class MsgHandlersRegistrator;
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/ReplicaBase.hpp]: Reducing the number of includes in header files by moving them to cpp files is usually a good practice, as it may cause a circular dependencies

File: bftengine/src/bftengine/BFTEngine.cpp [hunk 15]
```diff
@@ -167,14 +159,15 @@ Replica *Replica::createNewReplica(ReplicaConfig *replicaConfig,
...
+    replicaInternal->replica_.reset(new ReplicaImp(
+        loadedReplicaData, requestsHandler, stateTransfer, msgsCommunicatorPtr, persistentStoragePtr, msgHandlersPtr));
   }
-  preprocessor::PreProcessor::addNewPreProcessor(
-      msgsCommunicatorPtr, incomingMsgsStoragePtr, msgHandlersPtr, *requestsHandler, *replicaInternal->replica_);
+  preprocessor::PreProcessor::addNewPreProcessor(msgsCommunicatorPtr,
+                                                 incomingMsgsStoragePtr,
+                                                 msgHandlersPtr,
+                                                 *requestsHandler,
+                                                 *dynamic_cast<InternalReplicaApi *>(replicaInternal->replica_.get()));
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/BFTEngine.cpp]: replicaInternal->replica_ is of a type ReplicaBase, but in this case it's an instance of ReplicaImp.

File: bftengine/src/bftengine/ReplicaBase.hpp [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+  void SetAggregator(std::shared_ptr<concordMetrics::Aggregator> a) { metrics_.SetAggregator(a); }
+
+  virtual void start();
+  virtual void stop();
+  SeqNum getLastExecutedSequenceNum() const { return lastExecutedSeqNum; }
+  virtual bool isRunning() const;
+
+ protected:
+  // Message handling
+  virtual void onReportAboutInvalidMessage(MessageBase* msg, const char* reason) = 0;
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/ReplicaBase.hpp]: It's not always an exception's what() message

File: bftengine/src/bftengine/ReplicaBase.hpp [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+
+  virtual void send(MessageBase* m, NodeIdType dest) { sendRaw(m, dest); }
+  void sendToAllOtherReplicas(MessageBase* m) {
+    for (ReplicaId dest : repsInfo->idsOfPeerReplicas()) sendRaw(m, dest);
+  }
+  void sendRaw(MessageBase* m, NodeIdType dest);
+
+  bool validateMessage(MessageBase* msg) {
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/ReplicaBase.hpp]: bool validateMessage(MessageBase* msg) is not a virtual method - it's used to decide either to process message or to delete it.
virtual validate() for messages may throw and should be caught here.
virtual  onReportAboutInvalidMessage() is the way it was designed - the inheriting replica may decide to throw or even assert.

File: bftengine/src/bftengine/ReplicaBase.hpp [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+  // last SeqNum executed  by this replica (or its affect was transferred to this replica)
+  SeqNum lastExecutedSeqNum = 0;
+
+  //////////////////////////////////////////////////
+  // METRICS
+  concordMetrics::Component metrics_;
+
+  ///////////////////////////////////////////////////
+  // Timer manager/container
+  Timers timers_;
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/ReplicaBase.hpp]: You are right, the member should be removed.
As for the second part, as I understand Timers::evaluate() happens on the thread running 
IncomingMsgsStorageImp::dispatchMessages().
There's no locking in the TimersSingleton.

File: bftengine/src/bftengine/ReplicaForStateTransfer.cpp [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+  stateTransfer->startRunning(this);
+  ReplicaBase::start();  // msg communicator should be last in the starting chain
+}
+
+void ReplicaForStateTransfer::stop() {
+  TimersSingleton::getInstance().cancel(stateTranTimer_);
+  ReplicaBase::stop();
+}
+
+template <>
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/ReplicaForStateTransfer.cpp]: It's the syntax for partial template specialization

File: bftengine/src/bftengine/ReplicaForStateTransfer.hpp [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+  friend class MessageBase;
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/ReplicaForStateTransfer.hpp]: You are right, there's no need in this friend class declaration.

File: bftengine/src/bftengine/ReplicaBase.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (andrewjstone) [reply bftengine/src/bftengine/ReplicaBase.hpp]: It does, but it may operate on data structures running in another thread, in which case locks would be needed. The key point is that it centralizes mutable state.

File: bftengine/src/bftengine/ReplicaBase.hpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (andrewjstone) [reply bftengine/src/bftengine/ReplicaBase.hpp]: Ok. That makes it even *less* safe to use this interface then. The caller must guarantee a string is null terminated, and the callee must trust that implicitly. In general this type of interface is the #1 security hole in all C based software, and is why strcpy is generally recognized as unsafe. https://en.wikipedia.org/wiki/Buffer_overflow
